_posts/2019-07-16-neuron.md

```diff
@@ -0,0 +1,161 @@
+from the receptors. This sum is passed into some kind of activation function that essentially handles translating 
+that sum into a usable value, and then spits that value out of our neuron.
+```
+input 1
+     \
+   weight 1
+       \
+        +-----------+
+        | weighted  |     +----------+
+        |           | --> |activation| --> classification value
+        |    sum    |     +----------+
+        +-----------+
+       /
+   weight 2
+     /
+input 2
+```
+### Step one: creating the neuron class
+This is some pretty straightforward code just setting up our skeleton on which to build.
+```python
+class Neuron:
+    def __init__(self, number_of_inputs):
+        self.number_of_inputs = number_of_inputs
+
+n = Neuron(number_of_inputs=2)
+```
+In order to do some simple error checking down the road, I chose not to make input numbers dynamic. This will let me do
+things like:
+```python
+if len(inputs) != self.number_of_inputs:
+    raise NeuronException('Length of inputs must be the same as the declared number of inputs.')
+```
+It is also a handy counter for doing loops for weighted sums and training our model later. We know how many times we'll
+have to do a loop without having to grab the `len()` of our inputs or weights or whatever. And it is unlikely we 
+wouldn't know the dimensionality of our data before sending it to a neuron, so the slight _potential_ inconvenience of
+having to declare our size is offset by the gains in functionality.
+ 
+### Step two: weights
+Now we need weights. Can't do a weighted sum without weights!
+
+I want to be able to supply the neuron with a list of specific weights (for debugging or reproducing results for 
+plotting), or let it randomly generate weights on its own. We'll want these random weights to be normally distributed 
+around 0, `random.normalvariate(0, .5)` will be perfect for this.
+```python
+import random
+
+
+class Neuron:
+    def __init__(self, number_of_inputs, weights=None):
+        self.number_of_inputs = number_of_inputs
+        self.weights = weights if weights is not None else self._random_weights()
+
+    def _random_weights(self):
+        return [random.normalvariate(0, .5) for _ in range(self.number_of_inputs)]
+```
+This is another helpful use of having `number_of_inputs` explicit to the class; we don't have to wait until we see inputs
+for the first time to generate our random weights. We should also check that the supplied list of weights has the same
+number of entries as our expected number of inputs, so that every input has a weight. I'm omitting that code here for
+clarity.
+
+### Step three: inputs and sums
+Now that we have our weights, we also need to be able to take in a list of inputs and calculate our weighted sum.
+Again, I'm omitting the error checking for clarity.
+```python
+class Neuron:
+    # ...
+    def _weighted_sum(self, inputs):
+        return sum([inputs[i] * self.weights[i] for i in range(self.number_of_inputs)])
```

Comment:
For fun, could try out [pypy](https://pypy.org/) and see the speed difference

Reply:
That could be a fun "conclusion" to this series. run it normally and then run it in pypy and see what the speed changes are and compare them. I'll probably do that for funsies and then see if it makes a difference worth writing a post about.